File: Leetcode/Python3/92. Reverse Linked List II/solution.py
# https://leetcode.com/problems/reverse-linked-list-ii/


# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseBetween(self, head: Optional[ListNode], left: int, right: int) -> Optional[ListNode]:
        root = start = ListNode(0, head)
        for _ in range(left - 1):  # Set start index left - 1
            start = start.next
        end = start.next  # End node of reverse linked list
        for _ in range(right - left):
            temp, start.next, end.next = start.next, end.next, end.next.next
            start.next.next = temp
        return root

File: Leetcode/C++/92. Reverse Linked List II/solution.cpp
// https://leetcode.com/problems/reverse-linked-list-ii/description/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution
{
public:
    ListNode *reverseBetween(ListNode *head, int left, int right)
    {
        ListNode *root = new ListNode(0);
        ListNode *start = root;
        root->next = head;
        // During the swap, the values of start and end remain unchanged.
        for (int i = 0; i < left - 1; i++)
        {
            start = start->next;
        }
        ListNode *end = start->next;
        for (int i = 0; i < right - left; i++)
        {
            ListNode *temp = start->next;
            start->next = end->next;
            end->next = end->next->next;
            start->next->next = temp;
        }
        // The head node can be included in the range, so we create a dummy node and return its next pointer
        return root->next;
    }
};